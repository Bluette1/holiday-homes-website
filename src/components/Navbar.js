import Navbar from 'react-bootstrap/Navbar';
import Nav from 'react-bootstrap/Nav';
import NavDropdown from 'react-bootstrap/NavDropdown';
import Form from 'react-bootstrap/Form';
import FormControl from 'react-bootstrap/FormControl';
import Button from 'react-bootstrap/Button';
import CategoryFilter from './CategoryFilter';

const Header = () => (
  <Navbar bg="light" expand="lg">
    <Navbar.Brand href="#home">Holiday-Homes</Navbar.Brand>
    <Navbar.Toggle aria-controls="basic-navbar-nav" />
    <Navbar.Collapse id="basic-navbar-nav">
      <Nav className="mr-auto">
        <Nav.Link href="#home">Home</Nav.Link>
        <Nav.Link href="#link">
          <CategoryFilter />
        </Nav.Link>
        <NavDropdown title="Username" id="basic-nav-dropdown">
          <NavDropdown.Item href="#action/3.1">Profile</NavDropdown.Item>
          <NavDropdown.Item href="#action/3.2">Logout</NavDropdown.Item>
          <NavDropdown.Divider />
          <NavDropdown.Item href="#action/3.4">Favourites</NavDropdown.Item>
        </NavDropdown>
      </Nav>
      <Form inline>
        <FormControl type="text" placeholder="Search" className="mr-sm-2" />
        <Button variant="outline-success">Search</Button>
      </Form>
    </Navbar.Collapse>
  </Navbar>
);

export default Header;
