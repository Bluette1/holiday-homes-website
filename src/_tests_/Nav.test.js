import React from 'react';
import {
  render, screen, cleanup, waitFor, fireEvent,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import { ConnectedRouter } from 'connected-react-router';
import axios from 'axios';
import Navbar from '../components/Navbar';
import configureStore, { history } from '../store';
import { httpProtocol, host, port } from '../envVariables';

jest.mock('axios');

afterEach(cleanup);

const store = configureStore({
  user: { username: 'MaryS', authentication_token: 'token' },
});

const showFavouritesSpy = jest.fn();
const showNewHolidayHomeSpy = jest.fn();
const showDetailsSpy = jest.fn();
const showUserSpy = jest.fn();
const showSearchResultsSpy = jest.fn();
const NavbarWithStore = () => (
  <Provider store={store}>
    <React.StrictMode>
      <ConnectedRouter history={history}>
        <Navbar
          showFavourites={showFavouritesSpy}
          showNewHolidayHome={showNewHolidayHomeSpy}
          showDetails={showDetailsSpy}
          showUser={showUserSpy}
          showSearchResults={showSearchResultsSpy}
        />
      </ConnectedRouter>
    </React.StrictMode>
  </Provider>
);

test('renders the navbar - expected data is returned', async () => {
  axios.get.mockImplementation(url => {
    switch (url) {
      case `${httpProtocol}://${host}:${port}/categories`:
        return Promise.resolve({ data: { categories: [] } });
      default:
        return Promise.resolve({ data: [] });
    }
  });
  render(<NavbarWithStore />);
  await waitFor(() => {
    const categoryFilter = document.querySelector('.category-filter');
    expect(categoryFilter).toBeTruthy();
    const usernameBtn = screen.getByRole('button', { name: 'MaryS' });
    expect(usernameBtn).toBeInTheDocument();
    const homeBtn = screen.getByRole('button', { name: 'Home' });
    expect(homeBtn).toBeInTheDocument();
    const searchForm = screen.getByPlaceholderText('Search by title');
    const searchBtn = screen.getByRole('button', { name: 'Search' });
    expect(searchForm).toBeInTheDocument();
    expect(searchBtn).toBeInTheDocument();
  });
});

test('renders the navbar - expected data is returned and toggle buttons working correctly', async () => {
  axios.get.mockImplementation(url => {
    switch (url) {
      case `${httpProtocol}://${host}:${port}/categories`:
        return Promise.resolve({ data: { categories: [] } });
      default:
        return Promise.resolve({ data: [] });
    }
  });
  render(<NavbarWithStore />);
  await waitFor(() => {
    const categoryFilter = document.querySelector('.category-filter');
    expect(categoryFilter).toBeTruthy();
    const usernameBtn = screen.getByRole('button', { name: 'MaryS' });
    expect(usernameBtn).toBeInTheDocument();
    fireEvent.click(usernameBtn);
    expect(usernameBtn).toBeDefined();

    const profileItem = screen.getByRole('button', { name: 'Profile' });
    expect(profileItem).toBeInTheDocument();

    const addHolidayHomeItem = screen.getByRole('button', { name: 'Add a holiday home' });
    expect(addHolidayHomeItem).toBeInTheDocument();

    const logoutItem = screen.getByRole('button', { name: 'Logout' });
    expect(logoutItem).toBeInTheDocument();

    const favouritesItem = screen.getByRole('button', { name: 'Favourites' });
    expect(favouritesItem).toBeInTheDocument();

    expect(screen).toMatchSnapshot();

    const homeBtn = screen.getByRole('button', { name: 'Home' });
    expect(homeBtn).toBeInTheDocument();
    const searchForm = screen.getByPlaceholderText('Search by title');
    const searchBtn = screen.getByRole('button', { name: 'Search' });
    expect(searchForm).toBeInTheDocument();
    expect(searchBtn).toBeInTheDocument();
  });
});

test('renders the navbar - the expanded buttons function as expected when clicked', async () => {
  axios.get.mockImplementation(url => {
    switch (url) {
      case `${httpProtocol}://${host}:${port}/categories`:
        return Promise.resolve({ data: { categories: [] } });
      default:
        return Promise.resolve({ data: [] });
    }
  });

  render(<NavbarWithStore />);
  await waitFor(() => {
    const searchInput = screen.getByPlaceholderText('Search by title');
    const value = 'Courts';
    fireEvent.change(searchInput, { target: { value } });
    expect(searchInput.value).toBe(value);
    const searchBtn = screen.getByRole('button', { name: 'Search' });

    fireEvent.click(searchBtn);
    expect(showSearchResultsSpy).toHaveBeenCalled();
    expect(showSearchResultsSpy).toHaveBeenCalledWith(true, value);

    const usernameBtn = screen.getByRole('button', { name: 'MaryS' });
    fireEvent.click(usernameBtn);

    const profileItem = screen.getByRole('button', { name: 'Profile' });
    fireEvent.click(profileItem);
    expect(showUserSpy).toHaveBeenCalled();

    const addHolidayHomeItem = screen.getByRole('button', { name: 'Add a holiday home' });

    fireEvent.click(addHolidayHomeItem);
    expect(showNewHolidayHomeSpy).toHaveBeenCalled();

    const favouritesItem = screen.getByRole('button', { name: 'Favourites' });
    fireEvent.click(favouritesItem);
    expect(showFavouritesSpy).toHaveBeenCalled();
    expect(screen).toMatchSnapshot();
  });
});

test('renders the navbar - with the logout button working as expected', async () => {
  axios.get.mockImplementation(url => {
    switch (url) {
      case `${httpProtocol}://${host}:${port}/categories`:
        return Promise.resolve({ data: { categories: [] } });
      default:
        return Promise.resolve({ data: [] });
    }
  });

  render(<NavbarWithStore />);
  await waitFor(() => {
    const usernameBtn = screen.getByRole('button', { name: 'MaryS' });
    fireEvent.click(usernameBtn);
    const logoutItem = screen.getByRole('button', { name: 'Logout' });
    expect(logoutItem).toBeInTheDocument();
    expect(screen).toMatchSnapshot();
  });
});
