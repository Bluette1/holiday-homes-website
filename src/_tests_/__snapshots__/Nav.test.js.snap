// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the navbar - expected data is returned and toggle buttons working correctly 1`] = `
Object {
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`renders the navbar - the expanded buttons function as expected when clicked 1`] = `
Object {
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`renders the navbar - with the logout button working as expected 1`] = `
Object {
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;
